ksql-common/src/main/java/io/confluent/ksql/function/UdfFactory.java

```diff
@@ -98,17 +99,99 @@ public String toString() {
+  public KsqlFunction getFunction(final List<Schema> paramTypes) {
+    final List<FunctionParameter> params = mapToFunctionParameter(paramTypes, true);
+    final KsqlFunction function = functions.get(params);
     if (function != null) {
       return function;
     }
 
+    if (paramTypes.stream().anyMatch(Objects::isNull)) {
+      return functions.entrySet()
+          .stream()
+          .filter(entry -> checkParamsMatch(entry.getKey(), params))
+          .map(Map.Entry::getValue)
+          .findFirst().orElseThrow(() -> createNoMatchingFunctionException(paramTypes));
+    }
+
+    throw createNoMatchingFunctionException(paramTypes);
+  }
+
+  private KsqlException createNoMatchingFunctionException(final List<Schema> paramTypes) {
     final String sqlParamTypes = paramTypes.stream()
-        .map(SchemaUtil::getSchemaTypeAsSqlType)
+        .map(schema -> schema == null
+            ? null
+            : SchemaUtil.getSchemaTypeAsSqlType(schema.type()))
         .collect(Collectors.joining(", ", "[", "]"));
 
-    throw new KsqlException("Function '" + metadata.getName()
+    return new KsqlException("Function '" + metadata.getName()
                             + "' does not accept parameters of types:" + sqlParamTypes);
   }
+
+  private static boolean checkParamsMatch(final List<FunctionParameter> functionArgTypes,
+                                          final List<FunctionParameter> suppliedParamTypes) {
+    if (functionArgTypes.size() != suppliedParamTypes.size()) {
+      return false;
+    }
+
+    return IntStream.range(0, suppliedParamTypes.size())
+        .boxed()
+        .allMatch(idx -> suppliedParamTypes.get(idx).matches(functionArgTypes.get(idx)));
+  }
+
+  private List<FunctionParameter> mapToFunctionParameter(final List<Schema> params,
+                                                         final boolean includeNull) {
+    return params
+        .stream()
+        .filter(param -> includeNull || param != null)
+        .map(schema -> schema == null
+            ? new FunctionParameter(null, false)
+            : new FunctionParameter(schema.type(), schema.isOptional()))
+        .peek(fp -> {
+          if (!includeNull && fp.type == null) {
+            throw new KsqlException("Cannot add function with a null schema argument");
+          }
+        })
+        .collect(Collectors.toList());
+  }
+
+  private class FunctionParameter {
+    private final Schema.Type type;
+    private final boolean isOptional;
+
+    private FunctionParameter(final Schema.Type type, final boolean isOptional) {
+      this.type = type;
+      this.isOptional = isOptional;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      final FunctionParameter that = (FunctionParameter) o;
+      return type == that.type;
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
Personally - I'd document _why_ this does not include `isOptional` - same for `equals`. e.g.

```
  return Objects.hash(type /* isOptional intentionally excluded to... blah */);
```

And this from a man that does not like comments.